src/runtime/contrib/dnnl/dnnl_json_runtime.cc

```diff
@@ -389,13 +463,238 @@ class DNNLJSONRuntime : public JSONRuntimeBase {
+      dnnl::algorithm::pooling_avg_exclude_padding ;
+      
+    dnnl::memory::dim N = input_shape[0],
+      IC = input_shape[1],
+      IH = input_shape[2],
+      IW = input_shape[2],
+      KH = pool_size0,
+      KW = pool_size1,
+      PH_L = std::stoi(str_padding[1]),
+      PH_R = std::stoi(str_padding[3]),
+      PW_L = std::stoi(str_padding[0]),
+      PW_R = std::stoi(str_padding[2]),
+      SH = std::stoi(str_strides[0]),
+      SW = std::stoi(str_strides[1]),
+      DH = dilation0,
+      DW = dilation1;
+    
+    if(node.GetAttr<std::vector<std::string>>("layout")[0].size()>4)
+      {
+        IC = input_shape[1]*input_shape[4];                    // input channels
+    }
+
+    if(node.GetAttr<std::vector<std::string>>("layout")[0]=="NHWC")
+      {
+        IC = input_shape[3];
+        IH = input_shape[1];
+        IW = input_shape[2];
+    }
+
+
+    dnnl::memory::dim OH = (IH - KH + PH_L + PH_R) / SH + 1,
+                      OW = (IW - KW + PW_L + PW_R) / SW + 1;
+
+    // Memory shapes.
+    dnnl::memory::dims src_dims = {N, IC, IH, IW};
+    dnnl::memory::dims kernel_dims = {KH, KW};
+    dnnl::memory::dims dst_dims = {N, IC, OH, OW};
+    dnnl::memory::dims strides_dims = {SH, SW};
+    dnnl::memory::dims padding_dims_l = {PH_L, PW_L};
+    dnnl::memory::dims padding_dims_r = {PH_R, PW_R};
+    dnnl::memory::dims dilation = {DH, DW};
+
+    // Memory descriptions.
+    auto pool_src_md = dnnl::memory::desc(src_dims, dt::f32, src_df);
+    auto pool_dst_md = dnnl::memory::desc(dst_dims, dt::f32, dst_df); 
+
+    // AvgPool2d description.
+    auto avgpool_desc = dnnl::pooling_forward::desc(
+      dnnl::prop_kind::forward_inference, alg_kind, pool_src_md,
+      pool_dst_md, strides_dims, kernel_dims,
+      padding_dims_l, padding_dims_r
+    );
+
+    auto avgpool_prim_desc = dnnl::pooling_forward::primitive_desc(avgpool_desc, engine_, true);//allow_enpty=true
+
+    // Push to the network.
+    auto pool = dnnl::pooling_forward(avgpool_prim_desc);
+    net_.push_back(pool);
+
+    // Memories
+    // ICHECK_EQ(node.GetAttr<std::vector<std::string>>("layout")[0], "NCHW"); // naming
+    auto pool2d_src_memory = BindDNNLMemory(data_entry, pool_src_md);
+
+    JSONGraphNodeEntry out_entry(nid, 0);
+    auto pool2d_dst_memory = BindDNNLMemory(out_entry, avgpool_prim_desc.dst_desc());
+
+    // Bind memory buffers.
+    net_args_.push_back({
+      {DNNL_ARG_SRC, pool2d_src_memory},
+      {DNNL_ARG_DST, pool2d_dst_memory}
+    });
+  }
+
   // Read from DNNL memory (+offset) and write to the handle.
   inline void read_from_dnnl_memory(void* handle, const dnnl::memory& mem, size_t size,
                                     size_t offset = 0) {
-    uint8_t* src = static_cast<uint8_t*>(mem.get_data_handle());
-    std::copy(src + offset, src + offset + size, static_cast<uint8_t*>(handle));
+    mem.set_data_handle(handle);
   }
 
+  // Read from the handle and write to DNNL memory (+offset).
```

Comment:
Why do we comment out the non-copy version of `write_to_dnnl_memory`?

Reply:
Uncomment the write_to_dnnl_memory. It is never used.